### PR TITLE
Feat/injective

### PR DESCRIPTION
**Pull request by swiftcc624**
"injective front end GEN-211"

### CODE REVIEW

**Comment by parrot9design** [comment]
- [ ] The collateral list is currently alphabetical, so INJ should go after ICX.
Screenshot 2024-08-08 at 10 51 58 AM

---

- [ ] When you choose to swap INJ or deposit it as collateral, it defaults to the ICON blockchain. It should default to Injective.

- [ ] (This was missed in the Havah PR, but since it's related:) When you choose to swap HVH, it should default to Havah instead of the ICON blockchain.

---

- [ ] Let's tidy up the sign-in modal content:
ICON: `Supply liquidity. Vote.<br>Borrow, swap, & transfer cross-chain.`
Archway & Havah: `Swap & transfer cross-chain.`
EVM & Injective: `Borrow, swap, & transfer cross-chain.`

Screenshot 2024-08-08 at 11 40 01 AM

---

- [ ] Add the bridge fee for Injective (I have to assume it's not free?)

---

- [ ] Nothing happens when you click `Connect wallet` or `Choose address` for Injective:
Screenshot 2024-08-08 at 12 46 54 PM Screenshot 2024-08-08 at 12 46 44 PM

It should surface the Injective wallet modal instead:
Screenshot 2024-08-08 at 12 48 22 PM

---

- [ ] The UI can take a long time to register that an Injective transaction is complete.
- [ ] I've also had trouble getting funds onto Injective. I made a couple trades to get INJ and bnUSD, which were stuck in a pending state for a few minutes. The transactions have since been finalised, but nothing has appeared in my Injective wallet.

**Comment by swiftcc624** [comment]
@parrot9design 
> Add the bridge fee for Injective (I have to assume it's not free?)

It's showing fee, can you please give me a screenshot?

image

**Comment by parrot9design** [comment]
- [ ] The sign-in modal content for ICON should have a line break after `Vote`:

`Supply liquidity. Vote.<br>Borrow, swap, & transfer cross-chain.`

---

- [ ] I have 0.8 INJ and 2 bnUSD in my Injective wallet, but the app still doesn't detect my balances on Injective (if it helps, I'm using Keplr):

Screenshot 2024-08-12 at 10 47 59 AM Screenshot 2024-08-12 at 10 48 13 AM

Note that the INJ bridge fee is still missing for me, as well. Signed in, signed out, and in different browsers, it never appears.

**Comment by swiftcc624** [comment]
@parrot9design Can we remove the line break at the moment?

**Comment by swiftcc624** [comment]
@parrot9design It fetches balances correctly on my side.

image


Is there a way to test on different internet connections?

**Comment by parrot9design** [comment]
>Can we remove the line break at the moment?

I didn't expect a line break to cause any issues, but no worries. Let's just reverse the order: `Borrow, swap, & transfer cross-chain. Supply liquidity. Vote.`

>Is there a way to test on different internet connections?

Great call. Turned out an Injective API was getting blocked on our network, now that it's been allowed I can see my Injective balances. 🙌

**Comment by parrot9design** [comment]
- [ ] I'm unable to sign a transaction to deposit INJ collateral on Injective. I can't progress past the confirmation modal, even though I've left more than enough INJ to cover the fees.

I checked the console log, here's the error it reports when I click Deposit:
Screenshot 2024-08-12 at 3 30 54 PM


---

- [ ] It doesn't have to be resolved in this PR, but we seem to have different methods for rounding decimals:
Screenshot 2024-08-12 at 2 57 17 PM

**Comment by swiftcc624** [comment]
> I'm unable to sign a transaction to deposit INJ collateral on Injective. I can't progress past the confirmation modal, even though I've left more than enough INJ to cover the fees.

@hetfly Can you please complete InjectiveWalletXService.ts for collateral XTransactionTypes?

File: apps/web/src/app/pages/trade/bridge/_config/xChains.ts [hunk 1]
```diff
@@ -259,6 +259,31 @@ export const base: XChain = {
...
+  xChainType: 'INJECTIVE',
+  xWalletType: XWalletType.INJECTIVE,
+  tracker: 'https://explorer.injective.network/',
+  nativeCurrency: {
+    decimals: 18,
+    name: 'INJ',
+    symbol: 'INJ',
+  },
+  rpc: {
+    http: 'https://mainnet.base.org',
```
Comment (0xmilktea) [inline apps/web/src/app/pages/trade/bridge/_config/xChains.ts:275]: the url is wrong.

File: apps/web/src/app/pages/trade/bridge/_components/XTransferModal.tsx [hunk 2]
```diff
@@ -90,13 +94,23 @@ function XTransferModal() {
 
   const gasChecker = useXCallGasChecker(direction.from);
 
+  const ethereumChainId = useChainId();
```
Comment (0xmilktea) [inline apps/web/src/app/pages/trade/bridge/_components/XTransferModal.tsx:97]: use useAccount's chainId.

File: apps/web/src/app/pages/trade/xswap/_components/XSwapModal.tsx [hunk 3]
```diff
@@ -121,13 +124,23 @@ const XSwapModal = ({ account, currencies, executionTrade, direction, recipient,
 
   const gasChecker = useXCallGasChecker(direction.from);
 
+  const ethereumChainId = useChainId();
```
Comment (0xmilktea) [inline apps/web/src/app/pages/trade/xswap/_components/XSwapModal.tsx:127]: use useAccount's chainId

File: apps/web/src/packages/injective/InjectiveWalletOptionsModal.tsx [hunk 4]
```diff
@@ -0,0 +1,105 @@
+import React from 'react';
```
Comment (0xmilktea) [inline apps/web/src/packages/injective/InjectiveWalletOptionsModal.tsx:1]: move the file in app/components/WalletModal folder.

File: apps/web/src/packages/injective/InjectiveWalletOptionsModal.tsx [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+type InjectiveWalletOptionsModalProps = {
+  id?: MODAL_ID;
+};
+
+export const InjectiveWalletOptionsModal = ({
+  id = MODAL_ID.INJECTIVE_WALLET_OPTIONS_MODAL,
+}: InjectiveWalletOptionsModalProps) => {
+  const open = useModalStore(state => state.modals?.[id]);
+  const { connect } = useInjectiveWalletStore();
```
Comment (0xmilktea) [inline apps/web/src/packages/injective/InjectiveWalletOptionsModal.tsx:26]: useWalletModal hook

File: apps/web/src/packages/injective/InjectiveWalletOptionsModal.tsx [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+import { useInjectiveWalletStore } from '.';
+
+type InjectiveWalletOptionsModalProps = {
+  id?: MODAL_ID;
+};
+
+export const InjectiveWalletOptionsModal = ({
+  id = MODAL_ID.INJECTIVE_WALLET_OPTIONS_MODAL,
+}: InjectiveWalletOptionsModalProps) => {
+  const open = useModalStore(state => state.modals?.[id]);
```
Comment (0xmilktea) [inline apps/web/src/packages/injective/InjectiveWalletOptionsModal.tsx:25]: use useWalletModal hook

File: apps/web/src/packages/injective/InjectiveWalletOptionsModal.tsx [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+      name: 'Keplr',
+      wallet: Wallet.Keplr,
+      icon: KeplrIcon,
+    },
+  ];
+
+  const handleConnect = async (wallet: Wallet) => {
+    await connect(wallet);
+
+    // walletStrategy.setWallet(wallet);
```
Comment (0xmilktea) [inline apps/web/src/packages/injective/InjectiveWalletOptionsModal.tsx:46]: remove the dead code

File: apps/web/src/packages/injective/index.tsx [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+export const walletStrategy = new WalletStrategy({
+  chainId: ChainId.Mainnet,
+  ethereumOptions: {
+    ethereumChainId: EthereumChainId.Mainnet,
+    rpcUrl: mainnet.rpcUrls.default.http[0],
+  },
+});
+
+export const useInjectiveWalletStore = create<InjectiveWalletStore>()(
```
Comment (0xmilktea) [inline apps/web/src/packages/injective/index.tsx:26]: I personally prefer to use native React Context API for simple use case.

File: apps/web/src/app/components/CrossChainWalletConnect/index.tsx [hunk 9]
```diff
@@ -32,6 +33,8 @@ const CrossChainWalletConnect = ({ xChainId, editable }: { xChainId: XChainId; e
       connectKeplr();
     } else if (chain.xWalletType === XWalletType.HAVAH) {
       connectToHavah();
+    } else if (chain.xWalletType === XWalletType.INJECTIVE) {
+      modalActions.openModal(MODAL_ID.INJECTIVE_WALLET_OPTIONS_MODAL);
```
Comment (0xmilktea) [inline apps/web/src/app/components/CrossChainWalletConnect/index.tsx:37]: refactor it by using setWalletModal function.

File: apps/web/src/app/components/WalletModal/index.tsx [hunk 10]
```diff
@@ -150,6 +155,15 @@ export default function WalletModal() {
           xChains: undefined,
           switchChain: undefined,
         },
+        {
+          name: 'Injective',
+          logo: <InjectiveWalletIcon width="40" height="40" />,
+          connect: () => modalActions.openModal(MODAL_ID.INJECTIVE_WALLET_OPTIONS_MODAL),
```
Comment (0xmilktea) [inline apps/web/src/app/components/WalletModal/index.tsx:161]: refactor it by using setWalletMOdal

File: apps/web/package.json [hunk 11]
```diff
@@ -92,6 +97,7 @@
     "rlp": "3.0.0",
     "sanitize.css": "12.0.1",
     "serve": "11.3.2",
+    "stream-browserify": "3.0.0",
```
Comment (0xmilktea) [inline apps/web/package.json:100]: move it in dev deps

Comment (swiftcc624) [reply apps/web/package.json:100]: done

File: apps/web/src/packages/injective/InjectiveWalletOptionsModal.tsx (hunk 7 above)
Comment (swiftcc624) [reply apps/web/src/packages/injective/InjectiveWalletOptionsModal.tsx:46]: done

File: apps/web/src/app/pages/trade/bridge/_config/xChains.ts (hunk 1 above)
Comment (swiftcc624) [reply apps/web/src/app/pages/trade/bridge/_config/xChains.ts:275]: done

File: apps/web/src/app/pages/trade/bridge/_components/XTransferModal.tsx (hunk 2 above)
Comment (swiftcc624) [reply apps/web/src/app/pages/trade/bridge/_components/XTransferModal.tsx:97]: it doesn't work, the value is undefinend

File: apps/web/src/app/pages/trade/xswap/_components/XSwapModal.tsx (hunk 3 above)
Comment (swiftcc624) [reply apps/web/src/app/pages/trade/xswap/_components/XSwapModal.tsx:127]: it doesn't work, the value is undefinend

**Comment by parrot9design** [comment]
- [ ] ICX collateral is on the piss. It doesn't know whether I have collateral or not. This doesn't seem to be an issue in the live app.

https://github.com/user-attachments/assets/2c83225b-25ae-4ec9-b2a0-07cc0138badc

**Pull request merged**